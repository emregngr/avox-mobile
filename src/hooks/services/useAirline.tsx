import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirlines } from '@/services/airlineService'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirline() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirlinesData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirlines(selectedLocale),
    queryKey: ['allAirlines', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airlines: Airline[]): Airline[] => {
    if (!airlines) return []

    return airlines.filter(airline => {
      if (!airline) return false

      if (filters?.region) {
        const airlineRegion = airline?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airlineRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.businessType) {
        const businessType = airline?.operations?.businessType?.toLowerCase()
        const filterBusinessType = filters?.businessType?.toLowerCase()

        if (businessType !== filterBusinessType) {
          return false
        }
      }

      if (filters?.businessModel) {
        const businessModel = airline?.operations?.businessModel?.toLowerCase()
        const filterBusinessModel = filters?.businessModel?.toLowerCase()

        if (businessModel !== filterBusinessModel) {
          return false
        }
      }

      if (filters?.alliance) {
        const alliance = airline?.operations?.alliance?.toLowerCase()
        const filterAlliance = filters?.alliance?.toLowerCase()

        if (alliance !== filterAlliance) {
          return false
        }
      }

      if (filters?.minSkytraxRating && airline?.operations?.skytraxRating) {
        if (airline?.operations?.skytraxRating < filters?.minSkytraxRating) {
          return false
        }
      }

      const airlineRangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airline) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a.companyInfo?.foundingYear
              ? parseInt(String(a?.companyInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.companyInfo?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.companyInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.network?.destinationCount },
        { filterKey: 'destinationCountries', jsonPath: a => a?.network?.destinationCountries },
        { filterKey: 'domesticConnections', jsonPath: a => a?.network?.domesticConnections },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.network?.internationalConnections,
        },
        { filterKey: 'totalAirplane', jsonPath: a => a?.fleet?.totalAirplane },
        { filterKey: 'averageAgeYears', jsonPath: a => a?.fleet?.averageAgeYears },
        { filterKey: 'airplaneTypeCount', jsonPath: a => a?.fleet?.airplaneTypeCount },
      ]

      for (const check of airlineRangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airline)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirlines = useMemo(() => {
    if (!allAirlinesData) return []

    let filtered = allAirlinesData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airline =>
          airline.iataCode?.toLowerCase() === searchLower ||
          airline.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airline => {
          const name = airline?.name?.toLowerCase() || ''
          const city = airline?.operations?.hub?.city?.toLowerCase() || ''
          const country = airline?.operations?.country?.toLowerCase() || ''
          const iata = airline?.iataCode?.toLowerCase() || ''
          const icao = airline?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            icao.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirlinesData, debouncedSearch, filters])

  const paginatedAirlines = useMemo(() => {
    if (!filteredAirlines) return []
    return filteredAirlines.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirlines, page])

  const hasMore = paginatedAirlines.length < filteredAirlines.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirlines?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirlines,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
