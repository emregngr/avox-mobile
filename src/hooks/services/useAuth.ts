import { getApp } from '@react-native-firebase/app'
import type { FirebaseAuthTypes } from '@react-native-firebase/auth'
import { getAuth, getIdToken, onAuthStateChanged } from '@react-native-firebase/auth'
import type { QueryClient } from '@tanstack/react-query'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  handleLogout,
  sendPasswordResetLink,
  signInWithApple,
  signInWithEmail,
  signInWithGoogle,
  signUpWithEmail,
} from '@/services/authService'
import { login, logout, register, social } from '@/store/auth'
import useThemeStore from '@/store/theme'
import type { AuthCredentials, LoginCredentials, RegisterCredentials } from '@/types/feature/auth'
import { Logger } from '@/utils/common/logger'

const app = getApp()
const auth = getAuth(app)

const handleAuthSuccess = async <T extends AuthCredentials>(
  userCredential: FirebaseAuthTypes.UserCredential,
  authFunction: (data: T & { token: string }) => Promise<void>,
  authData: T,
  queryClient: QueryClient,
) => {
  try {
    const token = await getIdToken(userCredential.user, true)
    await authFunction({ ...authData, token })
    queryClient.invalidateQueries({ queryKey: ['user'] })
  } catch (error) {
    Logger.breadcrumb('Failed to handle auth success', 'error', error as Error)
  }
}

export const useUser = () =>
  useQuery({
    queryFn: () =>
      new Promise(resolve => {
        const unsubscribe = onAuthStateChanged(auth, userState => {
          unsubscribe()
          resolve(userState)
        })
      }),
    queryKey: ['user'],
    staleTime: Infinity,
  })

export const useEmailRegister = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, firstName, lastName, password }: RegisterCredentials) =>
      await signUpWithEmail({ email, firstName, lastName, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailRegisterFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        register,
        {
          email: variables.email,
          firstName: variables.firstName,
          lastName: variables.lastName,
          password: variables.password,
          token: '',
        },
        queryClient,
      )
    },
  })
}

export const useEmailLogin = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, password }: LoginCredentials) =>
      await signInWithEmail({ email, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        login,
        {
          email: variables.email,
          password: variables.password,
        },
        queryClient,
      )
    },
  })
}

const useSocialLogin = (
  signInMethod: () => Promise<FirebaseAuthTypes.UserCredential>,
  provider: 'google' | 'apple',
) => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await signInMethod(),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('socialLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async userCredential => {
      await handleAuthSuccess(userCredential, social, { provider }, queryClient)
    },
  })
}

export const useGoogleLogin = () => useSocialLogin(signInWithGoogle, 'google')

export const useAppleLogin = () => useSocialLogin(signInWithApple, 'apple')

export const useLogout = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await handleLogout(),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('logoutFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: async () => {
      await logout()
      queryClient.setQueryData(['user'], null)
      queryClient.removeQueries()
    },
  })
}

export const useForgotPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: sendPasswordResetLink,
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('passwordResetLinkSentMessageFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordResetLinkSentMessage'),
        type: 'success',
      })
      router.back()
    },
  })
}
