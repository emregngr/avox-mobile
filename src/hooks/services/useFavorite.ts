import { getApp } from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { useFocusEffect } from 'expo-router'
import { useCallback } from 'react'

import {
  addToFavorites,
  fetchFavoriteDetails,
  fetchFavoriteIds,
  removeFromFavorites,
} from '@/services/favoriteService'
import useLocaleStore from '@/store/locale'
import type { FavoriteItem } from '@/types/feature/favorite'

const app = getApp()
const auth = getAuth(app)

const QUERY_KEYS = {
  favoriteDetails: (userId: string, locale: string) => ['favoriteDetails', userId, locale],
  favorites: (userId: string) => ['favorites', userId],
} as const

const getCurrentUserId = () => auth?.currentUser?.uid as string

export const useFavoriteIds = (refetchOnFocus = false) => {
  const userId = getCurrentUserId()
  const queryClient = useQueryClient()

  const query = useQuery({
    enabled: !!userId,
    gcTime: 5 * 60 * 1000,
    queryFn: fetchFavoriteIds,
    queryKey: QUERY_KEYS?.favorites(userId),
    staleTime: 1 * 60 * 1000,
  })

  useFocusEffect(
    useCallback(() => {
      const getFavorites = async () => {
        if (refetchOnFocus && userId) {
          await queryClient.invalidateQueries({
            queryKey: QUERY_KEYS?.favorites(userId),
          })
        }
      }
      getFavorites()
    }, [refetchOnFocus, userId, queryClient]),
  )

  return query
}

export const useFavoriteDetails = (refetchOnFocus = false) => {
  const { selectedLocale } = useLocaleStore()
  const userId = getCurrentUserId()
  const queryClient = useQueryClient()
  const { data: favoriteIds = [] } = useFavoriteIds(refetchOnFocus)

  const query = useQuery({
    enabled: !!userId && !!selectedLocale,
    queryFn: async () => {
      if (!favoriteIds || favoriteIds?.length === 0) return Promise.resolve([])
      const favoriteDetails = await fetchFavoriteDetails(favoriteIds, selectedLocale)
      return favoriteDetails
    },
    queryKey: QUERY_KEYS?.favoriteDetails(userId, selectedLocale),
    staleTime: 1 * 60 * 1000,
  })

  useFocusEffect(
    useCallback(() => {
      const getFavoriteDetails = async () => {
        if (refetchOnFocus && userId) {
          await queryClient?.invalidateQueries({
            queryKey: QUERY_KEYS?.favoriteDetails(userId, selectedLocale),
          })
        }
      }
      getFavoriteDetails()
    }, [refetchOnFocus, userId, selectedLocale, queryClient, favoriteIds]),
  )

  return query
}

export const useAddFavorite = () => {
  const queryClient = useQueryClient()
  const userId = getCurrentUserId()
  const { selectedLocale } = useLocaleStore()

  return useMutation({
    mutationFn: addToFavorites,

    onError: (err, variables, context) => {
      if (context?.previousFavorites && userId) {
        queryClient?.setQueryData(QUERY_KEYS.favorites(userId), context?.previousFavorites)
      }
    },

    onMutate: async (variables: FavoriteItem) => {
      if (!userId) return

      await queryClient.cancelQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      const previousFavorites = queryClient?.getQueryData(QUERY_KEYS.favorites(userId))

      queryClient.setQueryData(QUERY_KEYS.favorites(userId), (old: any) => {
        if (!old) return [{ id: variables?.id, type: variables?.type }]

        const exists = old.some(
          (item: any) => item?.id === variables?.id && item?.type === variables?.type,
        )
        if (exists) return old
        return [...old, { id: variables?.id, type: variables?.type }]
      })

      return { previousFavorites }
    },

    onSuccess: async () => {
      if (!userId) return

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favoriteDetails(userId, selectedLocale),
      })
    },
  })
}

export const useRemoveFavorite = () => {
  const queryClient = useQueryClient()
  const userId = getCurrentUserId()
  const { selectedLocale } = useLocaleStore()

  return useMutation({
    mutationFn: removeFromFavorites,

    onError: (err, variables, context) => {
      if (context?.previousFavorites && userId) {
        queryClient?.setQueryData(QUERY_KEYS.favorites(userId), context?.previousFavorites)
      }
    },

    onMutate: async (variables: FavoriteItem) => {
      if (!userId) return

      await queryClient.cancelQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      const previousFavorites = queryClient.getQueryData(QUERY_KEYS.favorites(userId))

      queryClient.setQueryData(QUERY_KEYS.favorites(userId), (old: any) => {
        if (!old) return []
        return old?.filter(
          (item: any) => !(item?.id === variables?.id && item?.type === variables?.type),
        )
      })

      return { previousFavorites }
    },

    onSuccess: async () => {
      if (!userId) return

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favorites(userId),
      })

      await queryClient.invalidateQueries({
        queryKey: QUERY_KEYS.favoriteDetails(userId, selectedLocale),
      })
    },
  })
}

export const useIsFavorite = ({ id, type }: FavoriteItem) => {
  const { data: favoriteIds = [] } = useFavoriteIds()

  return favoriteIds?.some((fav: FavoriteItem) => fav?.id === id && fav?.type === type)
}
