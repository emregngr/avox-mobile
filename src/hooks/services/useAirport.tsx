import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirports } from '@/services/airportService'
import useLocaleStore from '@/store/locale'
import type { Airport } from '@/types/feature/airport'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirport() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirportsData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirports(selectedLocale),
    queryKey: ['allAirports', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airports: Airport[]): Airport[] => {
    if (!airports) return []

    return airports.filter(airport => {
      if (!airport) return false

      if (filters?.region) {
        const airportRegion = airport?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airportRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.airportType) {
        const airportType = airport?.operations?.airportType?.toLowerCase()
        const filterAirportType = filters?.airportType?.toLowerCase()

        if (airportType !== filterAirportType) {
          return false
        }
      }

      const airportServices = airport?.facilities?.services?.map(s => s?.toLowerCase()) || []

      const serviceKeywords: Record<string, string[]> = {
        hasCarRental: ['car rental'],
        hasChildrensPlayArea: ["children's play area"],
        hasDutyFree: ['duty-free shops'],
        hasHealthServices: ['health services'],
        hasHotels: ['hotels'],
        hasLounges: ['lounges'],
        hasPrayerRoom: ['prayer rooms'],
        hasRestaurants: ['restaurants'],
      }

      for (const serviceName in serviceKeywords) {
        if (filters?.[serviceName] === true) {
          const keywords = serviceKeywords?.[serviceName]

          const hasMatchingService = airportServices?.some(serviceString =>
            keywords?.some((keyword: string) => serviceString?.includes?.(keyword)),
          )

          if (!hasMatchingService) {
            return false
          }
        }
      }

      if (filters?.is24Hour === true && !airport?.operations?.is24Hour) return false
      if (filters?.is24Hour === false && airport?.operations?.is24Hour) return false

      if (filters?.hasMetro === true && !airport?.facilities?.hasMetro) return false
      if (filters?.hasMetro === false && airport?.facilities?.hasMetro) return false

      if (filters?.freeWifi === true && !airport?.facilities?.freeWifi) return false
      if (filters?.freeWifi === false && airport?.facilities?.freeWifi) return false

      if (filters?.minGoogleRating && airport?.facilities?.googleMapsRating) {
        if (airport?.facilities?.googleMapsRating < filters?.minGoogleRating) {
          return false
        }
      }

      const rangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airport) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a?.airportInfo?.foundingYear
              ? parseInt(String(a?.airportInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.infrastructure?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.airportInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.flightOperations?.destinationCount },
        {
          filterKey: 'destinationCountries',
          jsonPath: a => a.flightOperations?.destinationCountries,
        },
        {
          filterKey: 'domesticConnections',
          jsonPath: a => a?.flightOperations?.domesticConnections,
        },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.flightOperations?.internationalConnections,
        },
        { filterKey: 'elevationFt', jsonPath: a => a?.operations?.location?.elevationFt },
        {
          filterKey: 'passengerCapacity',
          jsonPath: a => a?.infrastructure?.passengerCapacity,
        },
        { filterKey: 'baggageCapacity', jsonPath: a => a?.infrastructure?.baggageCapacity },
        { filterKey: 'terminalCount', jsonPath: a => a?.infrastructure?.terminalCount },
        {
          filterKey: 'terminalAreaHectares',
          jsonPath: a => a?.infrastructure?.terminalAreaHectares,
        },
        { filterKey: 'airportAreaHectares', jsonPath: a => a?.infrastructure?.airportAreaHectares },
        { filterKey: 'apronCount', jsonPath: a => a?.infrastructure?.apronCount },
        { filterKey: 'runwayCount', jsonPath: a => a?.infrastructure?.runwayCount },
        { filterKey: 'mainRunwayLengthM', jsonPath: a => a?.infrastructure?.runways?.lengthM },
        { filterKey: 'towerHeightM', jsonPath: a => a?.infrastructure?.towerHeightM },
        {
          filterKey: 'parkingCapacityVehicles',
          jsonPath: a => a?.facilities?.parkingCapacityVehicles,
        },
        { filterKey: 'loungeCount', jsonPath: a => a?.facilities?.loungeCount },
        { filterKey: 'securityQueueTime', jsonPath: a => a?.facilities?.securityQueueTime },
        { filterKey: 'checkinTimeAvg', jsonPath: a => a?.facilities?.checkinTimeAvg },
      ]

      for (const check of rangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airport)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirports = useMemo(() => {
    if (!allAirportsData) return []
    let filtered = allAirportsData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airport =>
          airport.iataCode?.toLowerCase() === searchLower ||
          airport.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airport => {
          const name = airport?.name?.toLowerCase() || ''
          const city = airport?.operations?.location?.city?.toLowerCase() || ''
          const country = airport?.operations?.country?.toLowerCase() || ''
          const iata = airport?.iataCode?.toLowerCase() || ''
          const ident = airport?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            ident.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirportsData, debouncedSearch, filters])

  const paginatedAirports = useMemo(() => {
    if (!filteredAirports) return []
    return filteredAirports.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirports, page])

  const hasMore = paginatedAirports.length < filteredAirports.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirports?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirports,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
