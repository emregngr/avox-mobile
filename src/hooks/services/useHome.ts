import { useQuery } from '@tanstack/react-query'

import { getAllHomeData } from '@/services/homeService'
import useLocaleStore from '@/store/locale'

export function useHome() {
  const { selectedLocale } = useLocaleStore()

  const {
    data: homeData,
    error,
    isLoading,
  } = useQuery({
    queryFn: async () => await getAllHomeData(selectedLocale),
    queryKey: ['homeData', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  return {
    error,
    homeData,
    isLoading,
  }
}
