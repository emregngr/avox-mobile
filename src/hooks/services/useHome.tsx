import { useQuery } from '@tanstack/react-query'

import { getAllHomeData } from '@/services/homeService'
import useLocaleStore from '@/store/locale'

export function useHome() {
  const { selectedLocale } = useLocaleStore()

  const {
    data: homeData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllHomeData(selectedLocale),
    queryKey: ['home', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  return {
    error,
    homeData,
    isLoading,
  }
}
