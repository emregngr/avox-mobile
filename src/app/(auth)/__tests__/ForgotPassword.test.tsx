import { fireEvent, render, waitFor } from '@testing-library/react-native'
import type { ReactNode } from 'react'
import React from 'react'
import { useForm } from 'react-hook-form'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import ForgotPassword from '@/app/(auth)/forgot-password'
import { useForgotPassword } from '@/hooks/services/useAuth'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('react-hook-form')

const mockedUseForm = useForm as jest.MockedFunction<typeof useForm>

jest.mock('@/components/common', () => {
  const { View, Text, TouchableOpacity, TextInput } = require('react-native')

  return {
    Header: ({ title, backIconOnPress }: { title: string; backIconOnPress: () => void }) => (
      <>
        <Text testID="header-title">{title}</Text>
        <TouchableOpacity onPress={backIconOnPress} testID="back-button">
          <Text>Back</Text>
        </TouchableOpacity>
      </>
    ),

    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    TextInputField: ({ label, editable, ...props }: { label: string; editable: boolean }) => (
      <View>
        <Text>{label}</Text>
        <TextInput {...props} accessibilityLabel={label} editable={editable} />
      </View>
    ),

    ThemedGradientButton: ({
      onPress,
      label,
      disabled,
      loading,
    }: {
      onPress: () => void
      label: string
      disabled: boolean
      loading: boolean
    }) => (
      <TouchableOpacity disabled={disabled || loading} onPress={onPress} testID="submit-button">
        <Text>{loading ? 'Loading...' : label}</Text>
      </TouchableOpacity>
    ),
  }
})

const mockedForgotPassword = jest.fn()

jest.mock('@/hooks/services/useAuth')

const mockedUseForgotPassword = useForgotPassword as jest.MockedFunction<typeof useForgotPassword>

const mockedReset = jest.fn()

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseForgotPassword.mockReturnValue({
    isPending: false,
    mutateAsync: mockedForgotPassword,
  } as any)

  mockedUseForm.mockReturnValue({
    control: {},
    handleSubmit: jest.fn(callback => () => callback({ email: 'test@example.com' })),
    formState: { isValid: true },
    reset: mockedReset,
  } as any)
})

describe('ForgotPassword Screen', () => {
  it('should render correctly and button should be enabled', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<ForgotPassword />)
    const submitButton = getByTestId('submit-button')
    expect(submitButton).toBeTruthy()
    expect(submitButton).not.toBeDisabled()
  })

  it('should disable button when form is invalid after submission attempt', async () => {
    mockedUseForm.mockReturnValue({
      control: {},
      handleSubmit: jest.fn(callback => () => callback({ email: '' })),
      formState: { isValid: false },
      reset: mockedReset,
    } as any)
    const { getByTestId } = renderWithSafeAreaProvider(<ForgotPassword />)
    const submitButton = getByTestId('submit-button')
    expect(submitButton).not.toBeDisabled()
    fireEvent.press(submitButton)
    await waitFor(() => {
      expect(submitButton).toBeDisabled()
    })
  })

  it('should call forgotPassword and reset on submit', async () => {
    const handleSubmitMock = jest.fn(callback => () => callback({ email: 'test@example.com' }))
    mockedUseForm.mockReturnValue({
      control: {},
      handleSubmit: handleSubmitMock,
      formState: { isValid: true },
      reset: mockedReset,
    } as any)
    const { getByTestId } = renderWithSafeAreaProvider(<ForgotPassword />)
    const submitButton = getByTestId('submit-button')
    fireEvent.press(submitButton)
    expect(handleSubmitMock).toHaveBeenCalled()
    expect(mockedForgotPassword).toHaveBeenCalledWith('test@example.com')
    expect(mockedReset).toHaveBeenCalled()
  })

  it('should show loading state and disable button', () => {
    ;(require('@/hooks/services/useAuth').useForgotPassword as jest.Mock).mockReturnValue({
      isPending: true,
      mutateAsync: mockedForgotPassword,
    })

    const { getByTestId, getByText, getByLabelText } = renderWithSafeAreaProvider(
      <ForgotPassword />,
    )
    const submitButton = getByTestId('submit-button')

    expect(submitButton).toBeDisabled()
    expect(getByText('Loading...')).toBeTruthy()

    const emailInput = getByLabelText('email')
    expect(emailInput.props.editable).toBe(false)
  })
})

describe('ForgotPassword Screen Snapshot', () => {
  it('should render the ForgotPassword successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<ForgotPassword />)

    expect(toJSON()).toMatchSnapshot()
  })
})
