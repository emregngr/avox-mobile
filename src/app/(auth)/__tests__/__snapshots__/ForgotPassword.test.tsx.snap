// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ForgotPassword Screen Snapshot should render the ForgotPassword successfully 1`] = `
<View
  testID="safe-layout"
>
  <Text
    testID="header-title"
  >
    forgotPassword
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Text>
      Back
    </Text>
  </View>
  <RCTScrollView
    bottomOffset={50}
    contentContainerClassName="px-4"
    contentContainerStyle={
      {
        "paddingBottom": 20,
        "paddingTop": 64,
      }
    }
    keyboardDismissMode="on-drag"
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        className="gap-y-2"
      >
        <View>
          <Text>
            email
          </Text>
          <TextInput
            accessibilityLabel="email"
            control={{}}
            editable={true}
            keyboardType="email-address"
            name="email"
            onSubmitEditing={[Function]}
            placeholder="emailPlaceholder"
            returnKeyType="done"
            testID="email"
          />
        </View>
      </View>
      <View
        className="mt-6"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="submit-button"
        >
          <Text>
            sendResetLink
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
